src/bucket/BucketSnapshot.cpp

```diff
@@ -101,6 +103,20 @@ BucketSnapshot::loadKeys(std::set<LedgerKey, LedgerEntryIdCmp>& keys,
     auto indexIter = index.begin();
     while (currKeyIt != keys.end() && indexIter != index.end())
     {
+        if (lkMeter)
+        {
+            // If the transactions containing this key have a remaining
+            // quota less than the size of the key, we cannot load the entry.
+            auto keySize = xdr::xdr_size(*currKeyIt);
+            if (!lkMeter->canLoad(*currKeyIt, keySize))
+            {
+                // Consume key size bytes from the quotas of transactions
+                // containing the key (i.e. set their remaining quota to 0).
+                lkMeter->updateReadQuotasForKey(*currKeyIt, keySize);
```

Comment:
Sorry this is a bit confusing. We cannot know the entry size without reading, but in this case we know that we cannot read even the key with the quota (i.e. all transactions which contain this key have less than the key size remaining quota). The call to `updateReadQuota` zeros out the remaining quota and ensures we do not load subsequent keys which may be inside the remaining quota, as we know at this point the whole transaction will fail.

Reply:
To give an example: A transaction has 100 remaining quota, and two keys A (size=101), B (size=10). If we know that we cannot read the key of A, then we cannot read the entry of A. If B is only contained in this transaction, there is no point in reading B, even tho 10<100, as the overall transaction will fail.

Reply:
By consuming 100 from the meter, we will attempt to read B.

Reply:
Hmm, I see. That's a bit imprecise because keys shouldn't be counted towards read bytes and for a lot of contract entries keys may actually be much larger than the values. E.g. a balance key is at least about ~80 bytes, while the value is just ~10 bytes. The simplest fix is probably to just subtract key size from the entry size. I'm not sure if that's 100% accurate (because there can be a few additional bytes of overhead in `LedgerEntry` XDR), but that's also just a one-line fix. A bit more involved fix would be to have a separate function that sets quota to 0 when key can't be read, but doesn't subtract the key size from quota otherwise.

Reply:
> E.g. a balance key is at least about ~80 bytes, while the value is just ~10 bytes. 

Sure, but the entry we are reading would be 90 bytes, right? Or do we not account for key size in soroban?

Reply:
If I understand correctly, we charge for the entire `LedgerEntry` size, which always includes `LedgerKey`. In your example do we not charge 90 bytes for the balance entry @dmkozk?